Cypress based E2E tests

gabrieldutra opened the pull request:
## Description
PR for issue #3000.
Creates an initial infrastructure to run [Cypress](https://www.cypress.io/) tests. Also creates some initial example tests (Login, Logout and Create Data Source.

## Executing Cypress tests
### Locally
Note: `npm install` and `npm run build` must be updated

1. Run `npm run cypress:server start` - This will start docker, create the DB and run Redash's initial setup. If you've never run Redash through Docker before this might take a while.
2. Run `npm run cypress:open` (to open Cypress UI) or `npm run cypress:run` (to execute through command line). It's possible to pass parameters here by using `--` (e.g.: `npm run cypress:run -- -h`)
3. When finish testing, run `npm run cypress:server stop` to stop the server

You can also run a full `cypress:run` with `npm run cypress`

## External URL
The Base URL can be changed by using an env var:
`CYPRESS_baseUrl="http://externalredashurl" npm run cypress:open`

## Code Notes
- The recommended approach to select elements in the tests is to use `cy.contains('Text to be found')` or `data-*` attributes, I created those where I thought it was important in the tests.
- By creating only a separate server script we keep Cypress CLI features;
- So far the tests don't require a database reset, even though docker-compose makes sure it happens after a `npm run cypress:server stop`

## Cypress
![cypress](https://user-images.githubusercontent.com/3356951/47756038-50d69e80-dc7f-11e8-9d7b-3b97108e7d61.gif)

Please let me know if you have any questions, thoughts or improvement suggestions :grin:

arikfr commented:
12 minutes is brutal. Specially for CI. 

Maybe locally we will assume it’s installed and on the CI server we will use the Cypress Docker image?

arikfr commented:
As for CodeClimate -  maybe there is some eslint plug-in we can use (like the one for jest)? Otherwise we can tell it to ignore the cypress directory.

gabrieldutra commented:
**Installing time**
I looked at it more carefully and noticed that it took the 12 minutes only in netlify. With CircleCI it took less than a minute. In my dev environment I deleted its cache and it took 1 minute and 40 seconds to install. For me it's acceptable. I don't know why netlify took that much, but we can change its script to use `CYPRESS_INSTALL_BINARY=0 npm install` since we won't need cypress there.

I think it'll be perfect fine if we use circleci for Cypress purpose, it won't even need a separate docker image.

**CodeClimate**
So far I couldn't find a way of integrating the eslint plugin. The problem I'm facing is that CodeClimate only allows one config file for the whole project and if I try to configure the plugin in the existing `client/.eslintrc.js` what happens is a kind of mix up of settings. All cypress errors disappear, except `Missing semicolon` (which seems to be a good practice not to use with cypress). I'll keep looking for alternatives, but if you have any ideas, let me know :laughing:

arikfr commented:
> I actually didn't find anywhere specific saying it is supposed not to have them, it's just that they are never used in docs, examples, etc. Well, I don't have any concerns about just putting them 😁

I guess they didn't use them in the examples because they don't use them in their codebase, but not for any technical reasons. But in our codebase we do use semicolons, so let's just use them in the tests as well.

gabrieldutra commented:
I added `/* eslint-disable import/no-extraneous-dependencies, no-console */` to the top of the `cypress-server.js` Node script for the following:
- It was complaining about the `request` dependency, saying it should be in the project dependencies instead of devDependencies, which I think is wrong since Cypress is a devDependency
- I disabled console warnings in this file, which I think are good to inform what's happening.

jezdez commented:
@arikfr Speaking of CircleCI, do you want to enable GitHub checks for CircleCI? https://circleci.com/blog/see-the-status-of-your-circleci-workflows-in-github/

gabrieldutra commented:
Yes, @arikfr, I think what is left is only the integration. I think the easiest way to do this would be by adding something based on the existent `frontend-unit-tests`, but with step to install docker-compose and running `npm run cypress`

arikfr commented:
@jezdez done. 

@gabrieldutra yes. One issue you might run into is that you can't use volumes when using regular jobs (without `machine: true`). A way to avoid this is to use a different Docker Compose file (or extend the default one).

gabrieldutra commented on .circleci/config.yml:
```diff
@@ -54,6 +54,18 @@ jobs:
+      - setup_remote_docker
+      - checkout
+      - run: npm install
+      - run: npm build
```
I suppose we won't need both `npm install` and `npm build` then

gabrieldutra commented:
I noticed that Cypress binary was being installed inside docker-containers, so I took the approach you recommended before: I removed Cypress from `package.json` and added a command `npm run cypress:install`.

About CircleCI I'm now facing the issue of not being able to connect to `127.0.0.1:5000` after docker-compose setup. I'll continue on that by tomorrow.

arikfr commented:
> A solution I found to this was running Cypress from a docker container with its Docker Image. However, it took 16 minutes to build...

Which Docker image did you use?

gabrieldutra commented:
> > A solution I found to this was running Cypress from a docker container with its Docker Image. However, it took 16 minutes to build...
> 
> Which Docker image did you use?

I used `cypress/browsers:chrome67` for cypress and server based on `redash/base:latest`, as in the existing compose file.

arikfr commented:
Merged. 🚀
